File: Algorithm/BOJ/Graph/2252.줄세우기.java
import java.util.*;

// DFS

public class Main {
    static ArrayList<Integer>[] a;
    static void boolean[] check;
    static void go(int x) {
        check[x] = true;
        for(int y : a[x]) {
            if(check[y] == false) {
                go(y);
            }
        }
        System.out.print(x + " ");
    }

    public static void main(String args[]) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int m = sc.nextInt();
        a = new ArrayList[n+1];
        check = new boolean[n+1];
        for (int i=1; i<=n; i++) {
            a[i] = new ArrayList<>();
        }
        for (int i=0; i<m; i++) {
            int x = sc.nextInt();
            int y = sc.nextInt();
            a[y].add(x);
        }
        for (int i=1; i<=n; i++) {
            if (check[i] == false) {
                go(i);
            }
        }
        System.out.println();
    }
}






// BFS

import java.util.*;

public class Main {
    public static void main(String args[]) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int m = sc.nextInt();
        List<Integer>[] a = (List<Integer>[]) new List[n+1];
        for (int i=1; i<=n; i++) {
            a[i] = new ArrayList<Integer>();
        }
        int[] ind = new int[n+1];
        for (int i=0; i<m; i++) {
            int x = sc.nextInt();
            int y = sc.nextInt();
            a[x].add(y);
            ind[y] += 1;
        }
        Queue<Integer> q = new LinkedList<Integer>();
        for (int i=1; i<=n; i++) {
            if (ind[i] == 0) {
                q.add(i);
            }
        }
        while (!q.isEmpty()) {
            int x = q.remove();
            System.out.print(x + " ");
            for (int y : a[x]) {
                ind[y] -= 1;
                if (ind[y] == 0) {
                    q.add(y);
                }
            }
        }
        System.out.println();
    }
}

File: Algorithm/BOJ/Graph/DAG/2252.줄세우기.py
### DFS ###

def go(x, a, check):
    check[x] = True

    for xi in a[x]:
        if not check[xi]:
            go(xi, a, check)
    
    print(x)


n, m = map(int, input().split())

a = [[] for _ in range(n+1)]
check = [False]*(n+1)

for _ in range(m):
    x, y = map(int, input().split())
    a[y].append(x)

for i in range(n):
    if not check[i]:
        go(i, a, check)






### BFS ###

from collections import deque

n, m = map(int, input().split())

a = [[] for _ in range(n+1)]
ind = [0]*(n+1)
check = [False]*(n+1)

for _ in range(m):
    x, y = map(int, input().split())
    a[x].append(y)
    ind[y] += 1

q = deque()

for i in range(1,n+1):
    if not ind[i]:
        q.append(i)

while q:
    p = q.pop()
    print(p)

    for ai in a[p]:
        ind[ai] -= 1
        if ind[ai] == 0:
            q.append(ai)